src/QsCompiler/QirGeneration/DebugInfoManager.cs

```diff
@@ -0,0 +1,569 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Quantum.QIR.Emission;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Ubiquity.NET.Llvm;
+using Ubiquity.NET.Llvm.DebugInfo;
+using Ubiquity.NET.Llvm.Instructions;
+using Ubiquity.NET.Llvm.Types;
+using Ubiquity.NET.Llvm.Values;
+
+namespace Microsoft.Quantum.QsCompiler.QIR
+{
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+
+    /// <summary>
+    /// This class holds shared utility routines for generating QIR debug information.
+    /// It contains a reference to the GenerationContext that owns it because it needs
+    /// access to the same shared state represented by the GenerationContext.
+    /// </summary>
+    internal sealed class DebugInfoManager
+    {
+        private static class Config
+        {
+            /// <summary>
+            /// Whether or not to emit debug information during QIR generation
+            /// </summary>
+            public static bool EnableDebugSymbols { get; set; } = false;
+
+            /// <summary>
+            /// Dwarf version we are using for the debug info in the QIR generation
+            /// </summary>
+            public static uint DwarfVersion { get; } = 4;
+
+            /// <summary>
+            /// Title of the CodeView module flag for debug info
+            /// </summary>
+            public static string CodeviewName { get; } = "CodeView";
+
+            /// <summary>
+            /// CodeView version we are using for the debug info in the QIR generation
+            /// </summary>
+            public static uint CodeviewVersion { get; } = 1;
+
+            /// <summary>
+            /// The source language information for Dwarf.
+            /// For now, we are using the C interface. Ideally this would be a user defined language for Q#.
+            /// </summary>
+            public static SourceLanguage QSharpLanguage { get; } = SourceLanguage.C99;
```

Comment:
nit: For read-only properties that return a constant, I think it's better to use `=>` - the syntax is shorter and it avoids increasing the size of the class with constant fields.

```suggestion
            /// <summary>
            /// Dwarf version we are using for the debug info in the QIR generation
            /// </summary>
            public static uint DwarfVersion => 4;

            /// <summary>
            /// Title of the CodeView module flag for debug info
            /// </summary>
            public static string CodeviewName => "CodeView";

            /// <summary>
            /// CodeView version we are using for the debug info in the QIR generation
            /// </summary>
            public static uint CodeviewVersion => 1;

            /// <summary>
            /// The source language information for Dwarf.
            /// For now, we are using the C interface. Ideally this would be a user defined language for Q#.
            /// </summary>
            public static SourceLanguage QSharpLanguage => SourceLanguage.C99;
```